chaincode/ledger_db.go

```diff
@@ -93,17 +94,38 @@ func (db *LedgerDB) KeyExists(key string) (bool, error) {
 	return buff != nil, err
 }
 
+// LazyPut is equivalent to Put but calls put if and only if the new value differ from the cached state
+func (db *LedgerDB) LazyPut(key string, object interface{}) error {
+	buff, err := json.Marshal(object)
+	if err != nil {
+		return errors.Internal(err)
+	}
+
+	cachedValue, exists := db.getTransactionState(key)
+
+	if exists && bytes.Equal(buff, cachedValue) {
+		return nil
+	}
+
+	return db.put(key, buff)
+}
+
 // Put stores an object in the chaincode db, if the object already exists it is replaced
 func (db *LedgerDB) Put(key string, object interface{}) error {
 	buff, _ := json.Marshal(object)
 
-	if err := db.cc.PutState(key, buff); err != nil {
+	return db.put(key, buff)
+}
+
+// put stores a value in the chaincode db whether it already exists or not
+func (db *LedgerDB) put(key string, value []byte) error {
+	if err := db.cc.PutState(key, value); err != nil {
```

Comment:
The `put` and `LazyPut` methods add complexity that I think could be avoided by splitting `ComputePlan.Save` into two methods:

```go
// Save the whole compute plan
func (cp *ComputePlan) Save(db *LedgerDB, ID string) error {
	err := db.Put(ID, cp)
	if err != nil {
		return err
	}
	return cp.SaveState(db)
}

// Only save the state
func (cp *ComputePlan) SaveState(db *LedgerDB) error {
	return db.Put(cp.StateKey, cp.State)
}
```

Each caller can then decide if they want to save the whole compute plan - or only the state. This also makes the saving behavior more explicit.

Reply:
Well there is different topics here.
First the separation of the compute plan and its state:
The fact that both of them are save distinctively is, from my point of view, a storage optimization and should not be explicit outside this context. Therefore I was not comfortable having a specific method to save the state.
Secondly the `LazyPut`:
I agree it add complexity but I think in the case of big asset like a compute plan it' s a interesting optimization even if we save the state separately. But we can also integrate this mechanism in the ` Put` method and use it for every asset.

Reply:
Re: make `LazyPut` the default. Are there any other cases where we:

- get an asset from the ledger
- then try to save it without modifications ?

I can't think of any but I'm not familiar with all cases. If `LazyPut` would only be beneficial in the case of compute plans, then I'm not sure it's worth adding it since there's a simpler solution available. Moreover, `LazyPut` depends on the state of the cache, so future modifications of the cache system put it at risk of breakage (or worse, data corruption :scream:)

Regarding the two `Save` methods, we could also have a single `Save` method with an argument `stateOnly bool = false`, then we'd still only have a central place for saving that, by default, saves everything, but can be explicitly asked to only save the state. Would that work?

Reply:
You're right stating that it's not a common situation since update are really not the norm. In case of the computeplan it can happen easily and that's the whole reason behind this PR (that and the fact that a single PR should theoretically reach up to 25000 tuples and maybe more in the future). So let's drop my previous suggestion. (However I'm not overly worried about the cache since it's re-created at each transaction. The risk of data corruption is really not that high)
But my issue with your proposal `Save` and `SaveState` is not so much the two methods, it's having this separation of the state logic spreading to the tuples' code. But I'll implement it and we'll see. Let me prove myself wrong.

Reply:
To be clear - what I'm proposing is not to completely separate `Save` and `SaveSate`, it's to have two distinct methods but have `Save` call `SaveState`. This way `SaveState` can be called directly, but `Save` always saves the state too.